### ISSUE TITLE
Question on the distribution of test dataset

### ISSUE THREAD

**Issue by KyonQi** [SMILELab-FL/FedLab#293]
**Problem Description**

作者您好，我用`PartitionedCIFAR10`这个类划分一个IID的CIFAR10数据集，训练过程中我只设置了1个client。遇到问题是client模型在训练集上能够正常训练，loss正常下降，acc正常上升；但是在测试集上，acc始终只有0.1-0.2。

我实在不理解为什么会出现这种问题，作者大大能给我点帮助吗ORZ。

**Comment by AgentDS** [comment]
你能提供代码吗

**Comment by KyonQi** [comment]
代码结构大致如下：

我首先实例化了`PartitionedCIFAR10`
```
*****
train_transform = transforms.Compose([
            transforms.Pad(4),
            transforms.RandomHorizontalFlip(),
            transforms.RandomCrop(32),
            transforms.ToTensor(),
            transforms.Normalize((0.4914, 0.4822, 0.4465), (0.2023, 0.1994, 0.2010))
            ])

self.Partitioner = PartitionedCIFAR10(root="../../CIFAR10/", path="./splitData", dataname="cifar10", 
        num_clients=num_clients, download=False, preprocess=True, transform=train_transform)
******
```

测试集数据我是直接

```
transform_test = transforms.Compose([
            #transforms.transforms.Pad(2),
            transforms.ToTensor(),
            torchvision.transforms.Normalize((0.4914, 0.4822, 0.4465),(0.2023, 0.1994, 0.2010))
            ])
testset = torchvision.datasets.CIFAR10(root=self.root, train=False, download=False, transform=transform_test )
testloader = torch.utils.data.DataLoader(testset, batch_size=self.test_batch_size, shuffle=False, num_workers=4)

```


然后我利用这个实例化的`Partitioner.get_dataloader()`去获得Client的数据集，给到client的模型去训练。

```
*******
        for cid in range(self.num_clients): #client updates
            cid_trainloader = self.Partitioner.get_dataloader(cid, self.train_batch_size, type='train')  ######
            grads = self.client_train_one_iter(cid_trainloader, cid, lr, cur_iter)
*******
```

这个client_train_one_iter是把所有train_dataloader累计梯度，把梯度传回给server，让server做更新（我只设置有一个client，所以就直接更新了）。

```
def client_train_one_iter(self, train_loader, client_id, learning_rate, current_iter):
        grads = {'n_samples' : len(train_loader.dataset), 'named_grads' : {}}
        train_correct = 0
        train_loss = 0
        loss_func = torch.nn.CrossEntropyLoss()
        self._clients_model[client_id].train()
        self._clients_model[client_id].zero_grad()
        for iter, (imgs, targets) in enumerate(train_loader):
            imgs, targets = imgs.cuda(self.device), targets.cuda(self.device)
            y_preds = self._clients_model[client_id](imgs)
            loss = loss_func(y_preds, targets) / len(train_loader)
            train_loss += loss
            loss.backward()
            preds = y_preds.argmax(dim=1, keepdim=True)
            train_correct += preds.eq(targets.view_as(preds)).sum().item()
        print('[Client {:>2}]  Loss: {:>4.6f},  Accuracy: {:>.4f}'.format(client_id, train_loss, train_correct / len(train_loader.dataset)))
        for name, param in self._clients_model[client_id].named_parameters():
            grads['named_grads'][name] = param.grad.data
        return grads
```

每次做完参数更新，我把更新后的server模型直接丢给测试集看测试集的loss和acc，这里这个`evaluateTest`函数的`model_parameters`是更新完参数后的server参数，`testloader`就是最开始的测试集数据

```
def evaluateTest(self, model_parameters, testloader):
        test_loss = 0
        test_correct = 0
        loss_func = torch.nn.CrossEntropyLoss()
        self.test_model.load_state_dict(model_parameters)
        self.test_model.eval()
        with torch.no_grad():
            for iter, (imgs, targets) in enumerate(testloader):
                imgs, targets = imgs.cuda(self.device), targets.cuda(self.device)
                y_preds = self.test_model(imgs)
                loss = loss_func(y_preds, targets)
                test_loss += loss / len(testloader)
                preds = y_preds.argmax(dim=1, keepdim=True)
                test_correct += preds.eq(targets.view_as(preds)).sum().item()
        print('[Global Model]  Loss: {:>4.6f},  Accuracy: {:>.4f}'.format(test_loss, test_correct / len(testloader.dataset)))
        print('-----------')
        return test_loss, test_correct / len(testloader.dataset)
```

我也打印出来模型参数看过了，client模型和server模型参数是保持一致的，但是最后结果如下

```
[Client  0]  Loss: 7.148009,  Accuracy: 0.0004
[Global Model]  Loss: 5.995399,  Accuracy: 0.1091
-----------
[Client  0]  Loss: 5.423772,  Accuracy: 0.2123
[Global Model]  Loss: 6.189888,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 3.980572,  Accuracy: 0.2690
[Global Model]  Loss: 5.701934,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 3.891416,  Accuracy: 0.2135
[Global Model]  Loss: 5.996987,  Accuracy: 0.1231
-----------
[Client  0]  Loss: 3.458781,  Accuracy: 0.1910
[Global Model]  Loss: 5.397426,  Accuracy: 0.1005
-----------
[Client  0]  Loss: 3.298875,  Accuracy: 0.2016
[Global Model]  Loss: 5.124929,  Accuracy: 0.1003
-----------
[Client  0]  Loss: 2.815597,  Accuracy: 0.2402
[Global Model]  Loss: 4.389002,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 2.421301,  Accuracy: 0.2468
[Global Model]  Loss: 4.374753,  Accuracy: 0.1183
-----------
[Client  0]  Loss: 2.564128,  Accuracy: 0.2765
[Global Model]  Loss: 3.544712,  Accuracy: 0.1692
-----------
[Client  0]  Loss: 2.298079,  Accuracy: 0.2646
[Global Model]  Loss: 3.607571,  Accuracy: 0.1142
-----------
[Client  0]  Loss: 2.351645,  Accuracy: 0.2628
[Global Model]  Loss: 3.400596,  Accuracy: 0.1016
-----------
[Client  0]  Loss: 2.186417,  Accuracy: 0.2757
[Global Model]  Loss: 3.207944,  Accuracy: 0.1017
-----------
[Client  0]  Loss: 2.240705,  Accuracy: 0.2734
[Global Model]  Loss: 3.117150,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 2.014836,  Accuracy: 0.2985
[Global Model]  Loss: 3.108012,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 2.064108,  Accuracy: 0.2922
[Global Model]  Loss: 3.274139,  Accuracy: 0.1002
-----------
[Client  0]  Loss: 2.097387,  Accuracy: 0.3075
[Global Model]  Loss: 2.661020,  Accuracy: 0.1380
-----------
[Client  0]  Loss: 1.929211,  Accuracy: 0.3336
[Global Model]  Loss: 3.238064,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 2.035576,  Accuracy: 0.3168
[Global Model]  Loss: 2.637660,  Accuracy: 0.1439
-----------
[Client  0]  Loss: 1.894396,  Accuracy: 0.3343
[Global Model]  Loss: 3.175437,  Accuracy: 0.1001
-----------
[Client  0]  Loss: 1.857082,  Accuracy: 0.3504
[Global Model]  Loss: 2.588508,  Accuracy: 0.1009
-----------
[Client  0]  Loss: 1.935236,  Accuracy: 0.3357
[Global Model]  Loss: 3.609535,  Accuracy: 0.1000
-----------
[Client  0]  Loss: 1.807624,  Accuracy: 0.3558
[Global Model]  Loss: 2.866445,  Accuracy: 0.1005
-----------
******************
******************
```

在测试集上acc始终只有0.1左右，loss也降不下来。实在是不知道咋办好了ORZ。

**Comment by AgentDS** [comment]
首先，我想问一下
- 这个`def evaluateTest(self, model_parameters, testloader)`中的`model_parameter`是怎么获取的？
- 用的base model是什么？包含BatchNorm吗？如果是有BatchNorm的模型，你直接
```python
for name, param in self._clients_model[client_id].named_parameters():
       grads['named_grads'][name] = param.grad.data
```
这样更新的梯度不包含BatchNorm层的参数以及一些shortcut参数（这些都是untrainable parameter）。因此你的server model的BN参数始终是初始化的状态，因此输出的accuracy大概在10%左右

**Comment by AgentDS** [comment]
你可以打印输出`model.state_dict()`和`model.parameters()`看看是否有差异

**Comment by KyonQi** [comment]
我使用的是resnet18，我前面尝试过使用更简单的Lenet，确实就没遇到这方面的问题ORZ，我大概明白问题在哪了。

那请问有什么好方法能更新这类untrainable parameter吗？如果我只想通过梯度更新，是不是就不适用于这类带BN层的模型了？

**Comment by Zengdun-cs** [comment]
联邦学习中，batchnorm层的参数也需要聚合更新，这部分参数通常不会被`model.named_parameters()`和`model.parameters()`获取。针对这个问题，我们已经解决了之前的实现的问题，见[代码](https://github.com/SMILELab-FL/FedLab/blob/master/fedlab/utils/serialization.py)。如果需要使用这个版本可以从源码import一下。

**Comment by AgentDS** [comment]
直接用`model.state_dict()`得到的参数去global model进行参数更新就可以了，你训练可以直接在本地，然后直接上传训练好的local model完整的参数，然后在server部分执行`global_model.load_state_model(local_model.state_dict())`，类似这种。或者你可以直接用我们这边提供的模型参数序列化的工具去做参数copy。[参考这个函数](https://github.com/SMILELab-FL/FedLab/blob/be4737045c5964bef0f89931f2644c8bed8d18f5/fedlab/utils/serialization.py#L93)
